### PR TITLE
Migrate HTTP/1 and HTTP/2 to the new headers/cookies API

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:

We ported over a simpler API for HTTP headers and cookies from ServiceTalk, but we didn't put that API to use.
This PR rectifies this.

Modification:

- HTTP/1 has been changed to use the new headers API.
- HTTP/2 has also been changed to use the new headers API.
- The old HTTP/1 and HTTP/2 headers APIs have been removed, but the generic old headers API remain because it still has a number of niche uses.
- Tests have been updated.
- Two tests have been disabled; one appears to have a chaotic relationship with header validation, requiring it in some places but not in others, in the same pipeline; another relied on header iteration order being stable, to verify HPACK encoding against binary images.

Result:

Cleaner headers and cookies API for HTTP/1 and HTTP/2.

Draft for now because I think there's an issue with HPACK encoding, where some headers are supposed to be filtered out (!?) but aren't.

### CODE REVIEW

File: codec-http/src/main/java/io/netty5/handler/codec/http/HttpObjectDecoder.java [hunk 1]
```diff
@@ -609,25 +611,26 @@ private State readHeaders(Buffer buffer) {
...
-        List<String> contentLengthFields = headers.getAll(HttpHeaderNames.CONTENT_LENGTH);
-        if (!contentLengthFields.isEmpty()) {
+        Iterator<CharSequence> contentLengthFields = headers.valuesIterator(HttpHeaderNames.CONTENT_LENGTH);
+        boolean hasContentLength = contentLengthFields.hasNext();
+        if (hasContentLength) {
             HttpVersion version = message.protocolVersion();
-            boolean isHttp10OrEarlier = version.majorVersion() < 1 || (version.majorVersion() == 1
-                    && version.minorVersion() == 0);
+            boolean isHttp10OrEarlier = version.majorVersion() < 1 || version.majorVersion() == 1
+                                                                      && version.minorVersion() == 0;
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/HttpObjectDecoder.java:619]: I know the extra `(...)` is not needed but I still think it makes it easier to read..

File: codec-http/src/main/java/io/netty5/handler/codec/http/cors/CorsHandler.java [hunk 2]
```diff
@@ -226,6 +236,14 @@ private void setAllowPrivateNetwork(final HttpResponse response) {
         }
     }
 
+    private static CharSequence toCsv(Iterable<?> headerValues) {
+        StringJoiner joiner = new StringJoiner(",");
+        for (Object value : headerValues) {
+            joiner.add(StringUtil.escapeCsv(String.valueOf(value), true));
+        }
+        return joiner.toString();
+    }
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/cors/CorsHandler.java:245]:
```suggestion
    private static CharSequence toCsv(Iterable<?> headerValues) {
        Iterator<?> iterator = headerValues.iterator();
        if (!iterator.hasNext()) {
            return AsciiString.EMPTY_STRING;
        }
        StringJoiner joiner = new StringJoiner(",");
        do {
            Object value = iterator.next();
            joiner.add(StringUtil.escapeCsv(String.valueOf(value), true));
        }
        return joiner.toString();
    }
```

File: codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpHeaders.java [hunk 3]
```diff
@@ -81,9 +82,52 @@ public DefaultHttpHeaders(final int arraySizeHint, final boolean validateNames,
...
+            if (contentEqualsIgnoreCase(trim(rawNext.subSequence(begin, end)), expected)) {
+                return true;
+            }
+            begin = end + 1;
+        } while ((end = AsciiString.indexOf(rawNext, ',', begin)) != -1);
+
+        if (begin < rawNext.length()) {
+            return contentEqualsIgnoreCase(trim(rawNext.subSequence(begin, rawNext.length())), expected);
+        }
+        return false;
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpHeaders.java:130]: nit: you could share some code here by having the method take another boolean which influence if either `contentEquals` or `contentEqualsIgnoreCase` is used.

File: codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpHeaders.java [hunk 4]
```diff
@@ -731,7 +783,8 @@ private boolean remove(final CharSequence name, final CharSequence value, final
 
     @Override
     public String toString() {
-        return toString(DEFAULT_HEADER_FILTER);
+//        return toString(DEFAULT_HEADER_FILTER);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpHeaders.java:786]: Leftover ?
```suggestion
```

File: codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpSetCookie.java [hunk 5]
```diff
@@ -196,28 +202,35 @@ static HttpSetCookie parseSetCookie(final CharSequence setCookieString, boolean
...
-                    // Increment by 4 because we are skipping %0x##
-                    i += 4;
-                    break;
+//                case '%':
+//                    if (validateContent) {
+//                        extractAndValidateCookieHexValue(setCookieString, i);
+//                    }
+//                    // Increment by 4 because we are skipping %0x##
+//                    i += 4;
+//                    break;
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpSetCookie.java:230]: remove ?

File: codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpSetCookie.java [hunk 6]
```diff
@@ -196,28 +202,35 @@ static HttpSetCookie parseSetCookie(final CharSequence setCookieString, boolean
                         if (isWrapped) {
                             parseState = ParseState.Unknown;
                             value = setCookieString.subSequence(begin, i);
-                            // Increment by 3 because we are skipping DQUOTE SEMI SP
-                            i += 3;
+                            if (validateContent) {
+                                // Increment by 3 because we are skipping DQUOTE SEMI SP
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpSetCookie.java:206]: Can you link to the RFC ?

File: codec-http/src/main/java/io/netty5/handler/codec/http/headers/HeaderUtils.java [hunk 7]
```diff
@@ -512,6 +512,23 @@ private HttpCookiePair findNext(CharSequence cookieHeaderValue) {
             }
             return next;
         }
+
+        private int nextDelimiter(CharSequence cookieHeaderValue) {
+            boolean inQuotes = false;
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/headers/HeaderUtils.java:517]: Should we add some fast path for `AsciiString` that uses `forEachByte(...)` ?

File: codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/WebSocketUtil.java [hunk 8]
```diff
@@ -81,8 +81,8 @@ static byte[] randomBytes(int size) {
 
     static String calculateV13Accept(String nonce) {
         String concat = nonce + V13_ACCEPT_GUID;
-        byte[] sha1 = WebSocketUtil.sha1(concat.getBytes(CharsetUtil.US_ASCII));
-        return WebSocketUtil.base64(sha1);
+        byte[] sha1 = sha1(concat.getBytes(CharsetUtil.US_ASCII));
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/WebSocketUtil.java:84]:
```suggestion
        byte[] sha1 = sha1(concat.getBytes(StandardCharsets.US_ASCII));
```

File: codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java [hunk 9]
```diff
@@ -132,9 +132,7 @@ public Future<Void> write(final ChannelHandlerContext ctx, Object msg) {
                             }
                         };
 
-                        if (newHeaderValue != null) {
-                            headers.set(HttpHeaderNames.SEC_WEBSOCKET_EXTENSIONS, newHeaderValue);
-                        }
+                        headers.set(HttpHeaderNames.SEC_WEBSOCKET_EXTENSIONS, newHeaderValue);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java:135]: This is a behaviour change... just wanted to highlight this.

File: codec-http2/src/main/java/io/netty5/handler/codec/http2/Http2Exception.java [hunk 10]
```diff
@@ -29,7 +29,7 @@
  * Exception thrown when an HTTP/2 error was encountered.
  */
 @UnstableApi
-public class Http2Exception extends Exception {
+public class Http2Exception extends RuntimeException {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty5/handler/codec/http2/Http2Exception.java:32]: This change seems to be unrelated and maybe not what we want

File: codec-http2/src/main/java/io/netty5/handler/codec/http2/HttpToHttp2ConnectionHandler.java [hunk 11]
```diff
@@ -78,8 +78,11 @@ protected HttpToHttp2ConnectionHandler(Http2ConnectionDecoder decoder, Http2Conn
      * @throws Exception If the {@code httpHeaders} object specifies an invalid stream id
      */
     private int getStreamId(HttpHeaders httpHeaders) throws Exception {
-        return httpHeaders.getInt(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(),
-                                  connection().local().incrementAndGetNextStreamId());
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty5/handler/codec/http2/HttpToHttp2ConnectionHandler.java:82]: wow... this looks like a "bug" before as we always incremented the stream id even tho there might be the header. Let me open a PR against 4.1

File: codec-http2/src/main/java/io/netty5/handler/codec/http2/InboundHttpToHttp2Adapter.java [hunk 12]
```diff
@@ -37,8 +38,11 @@ public InboundHttpToHttp2Adapter(Http2Connection connection, Http2FrameListener
     }
 
     private static int getStreamId(Http2Connection connection, HttpHeaders httpHeaders) {
-        return httpHeaders.getInt(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(),
-                                  connection.remote().incrementAndGetNextStreamId());
+        CharSequence streamId = httpHeaders.get(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text());
+        if (streamId != null) {
+            return Integer.parseInt(streamId.toString());
+        }
+        return connection.remote().incrementAndGetNextStreamId();
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty5/handler/codec/http2/InboundHttpToHttp2Adapter.java:45]: same comment as in. https://github.com/netty/netty/pull/12735/files#diff-56499daa8bd2f57ca861aac556912e6f4d0d796694c9d4cb8e31c542cff7b6eeR80. I think this needs to be fixed in  4.1

File: common/src/main/java/io/netty5/util/AsciiString.java [hunk 13]
```diff
@@ -993,10 +993,10 @@ public AsciiString toUpperCase() {
      * @return a new string with characters {@code <= \\u0020} removed from the beginning and the end.
      */
     public static CharSequence trim(CharSequence c) {
-        if (c instanceof AsciiString) {
+        if (c.getClass() == AsciiString.class) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty5/util/AsciiString.java:996]: why this change ? Because of https://github.com/netty/netty/pull/12709 ?

File: common/src/main/java/io/netty5/util/AsciiString.java [hunk 14]
```diff
@@ -1533,6 +1533,28 @@ public static boolean contentEquals(CharSequence a, CharSequence b) {
...
+     * @param sequence The sequence to extract from.
+     * @param start The start index, inclusive.
+     * @param end The end index, exclusive.
+     * @return The subsequence.
+     */
+    public static CharSequence substring(CharSequence sequence, int start, int end) {
+        if (sequence.getClass() == AsciiString.class) {
+            return ((AsciiString) sequence).subSequence(start, end);
+        }
+        String str;
```
Comment (normanmaurer) [inline common/src/main/java/io/netty5/util/AsciiString.java:1549]:
```suggestion
        final String str;
```

File: example/src/main/java/io/netty5/example/http/snoop/HttpSnoopClient.java [hunk 15]
```diff
@@ -90,11 +89,9 @@ public static void main(String[] args) throws Exception {
...
 
             // Set some example cookies.
-            request.headers().set(
-                    HttpHeaderNames.COOKIE,
-                    ClientCookieEncoder.STRICT.encode(
-                            new DefaultCookie("my-cookie", "foo"),
-                            new DefaultCookie("another-cookie", "bar")));
+            request.headers()
+                   .addCookie(new DefaultHttpCookiePair("my-cookie", "foo"))
+                    .addCookie(new DefaultHttpCookiePair("another-cookie", "bar"));
```
Comment (normanmaurer) [inline example/src/main/java/io/netty5/example/http/snoop/HttpSnoopClient.java:94]:
```suggestion
                   .addCookie(new DefaultHttpCookiePair("another-cookie", "bar"));
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
First review done...

**Comment by slandelle** [comment]
Hi there,
Do you have some performance numbers to share to compare the new Cookie API vs the old one wrt parsing and serialization?

File: codec-http2/src/main/java/io/netty5/handler/codec/http2/HttpToHttp2ConnectionHandler.java (hunk 11 above)
Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty5/handler/codec/http2/HttpToHttp2ConnectionHandler.java:82]: actually never mind... this works as expected.

File: common/src/main/java/io/netty5/util/AsciiString.java (hunk 13 above)
Comment (franz1981) [reply common/src/main/java/io/netty5/util/AsciiString.java:996]: It shouldn't be a problem (not for the same reasons of #12709) because `AsciiString` is a concrete type which won't interact with the secondary cache field on Klass, see http://cr.openjdk.java.net/~jrose/jvm/checktype-2001.txt for more info how it works (with examples).

File: codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpSetCookie.java (hunk 6 above)
Comment (chrisvest) [reply codec-http/src/main/java/io/netty5/handler/codec/http/headers/DefaultHttpSetCookie.java:206]: It's https://www.rfc-editor.org/rfc/rfc6265#section-4.1.1
It'll add this to the comment.

File: codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java (hunk 9 above)
Comment (chrisvest) [reply codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java:135]: The `newHeaderValue` comes from `WebSocketExtensionUtil.computeMergeExtensionsHeaderValue` which never returns `null`.

File: codec-http2/src/main/java/io/netty5/handler/codec/http2/Http2Exception.java (hunk 10 above)
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty5/handler/codec/http2/Http2Exception.java:32]: Yeah, this change is because `DefaultHttp2Headers.validateKeys` do not declare any checked exceptions. Previously we used a sneaky-throw to bypass the compile-time check for this. Making this a `RuntimeException` is a pretty blunt hammer.
The alternative would be to throw `IllegalArgumentException` (or a new class derived from it) like the `MultiMap` derived headers used to do, and then catch that in all the places where we build header objects.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http/src/main/java/io/netty5/handler/codec/http/headers/HttpHeaders.java [hunk 16]
```diff
@@ -47,28 +47,54 @@
...
     static HttpHeaders newHeaders(int sizeHint, boolean checkNames, boolean checkCookies, boolean checkValues) {
         return new DefaultHttpHeaders(sizeHint, checkNames, checkCookies, checkValues);
     }
 
+    /**
+     * Create an independent copy of these HTTP headers.
+     *
+     * @return A new headers instance, with the same contents as this one.
+     */
+    HttpHeaders copy();
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http/src/main/java/io/netty5/handler/codec/http/headers/HttpHeaders.java:96]: A default method wouldn't know what concrete type to use.

File: codec-http/src/main/java/io/netty5/handler/codec/http/headers/HeaderUtils.java [hunk 17]
```diff
@@ -512,6 +514,44 @@ private HttpCookiePair findNext(CharSequence cookieHeaderValue) {
             }
             return next;
         }
+
+        private int nextDelimiterAscii(AsciiString cookieHeaderValue) {
+            ByteProcessor visitor = new ByteProcessor() {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http/src/main/java/io/netty5/handler/codec/http/headers/HeaderUtils.java]: I made the iterator itself be the byte processor. In the exposed API, the class itself is hiding behind an interface anyway so I think I can get away with that.

**Comment by chrisvest** [comment]
I ran the `HeaderBenchmark`, but only for 30 headers to keep the run-time reasonable.
Will need to make a new benchmark for cookie parsing, since the API is so different.

Iterating HTTP/2 headers is worse. The rest is improved or on-par:
Screen Shot 2022-08-26 at 17 10 50

**Comment by chrisvest** [comment]
@slandelle I brought back and migrated `ClientCookieDecoderBenchmark` to this new API. With the old parsing code I was getting `2.690 ± 0.024  ops/us`, and with a few adjustments I got the new code on par with `2.715 ± 0.208  ops/us`. A bit more variability, though.

**Comment by chrisvest** [comment]
All comments addressed. Please take a look, guys.

**Comment by chrisvest** [comment]
@normanmaurer Can I get another review?

File: common/src/main/java/io/netty5/util/AsciiString.java [hunk 18]
```diff
@@ -1547,6 +1544,28 @@ public static boolean contentEquals(CharSequence a, CharSequence b) {
...
+     * @param end The end index, exclusive.
+     * @return The subsequence.
+     */
+    public static CharSequence substring(CharSequence sequence, int start, int end) {
+        if (sequence.getClass() == AsciiString.class) {
+            return ((AsciiString) sequence).subSequence(start, end);
+        }
+        final String str;
+        if (sequence.getClass() == String.class) {
+            str = (String) sequence;
```
Comment (normanmaurer) [inline common/src/main/java/io/netty5/util/AsciiString.java:1562]: nit: I think calling `toString()` is good enough in both cases as if its a `String` it should be a NOOP.

**Review by normanmaurer — APPROVED** [review]
Just one nit
